service/history/mutableStatePubSub.go

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/pborman/uuid"
+	"github.com/uber/cadence/common/persistence"
+)
+
+type (
+	// MutableStatePubSub is used to get updates on mutable state changes.
+	MutableStatePubSub interface {
+		Publish(*persistence.WorkflowMutableState)
+		Subscribe() (string, <-chan struct{})
+		Unsubscribe(string)
+		LatestMutableState() *persistence.WorkflowMutableState
+	}
+
+	mutableStatePubSub struct {
+		// An atomic is used to keep track of mutable state rather than pushing updates on a channel.
+		// If a channel (buffered or unbuffered) is used, then at some point either blocking will occur or a mutable state update will be dropped.
+		// Since clients only care about getting the latest mutable state, simplify notifying clients
+		// of updates and providing the ability to get the latest satisfies all current use cases.
+		latestMS atomic.Value
+		lockableSubscribers
+	}
+
+	lockableSubscribers struct {
+		sync.RWMutex
+		subscribers map[string]chan struct{}
+	}
+)
+
+func (ls *lockableSubscribers) add() (string, chan struct{}) {
+	ls.Lock()
+	defer ls.Unlock()
+	id := uuid.New()
+	ch := make(chan struct{}, 1)
+	ls.subscribers[id] = ch
```

Comment:
does this id need to be unique across workflowExecutionContexts for the same workflow ? If not, consider using a map[int]chan struct{} - it likely will be faster and result in less allocations. The integer keys can be generated using a simple atomic 64-bit increment

Reply:
I will use int keys instead of uuid. But for now I will reuse the same Mutex primitive used for subscribers rather than introducing a new atomic concurrency primitive for this.